gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -11,118 +11,158 @@
-  return nonlinearGraph_.equals(other.nonlinearGraph_, tol) &&
+  return Base::equals(other, tol) &&
+         nonlinearGraph_.equals(other.nonlinearGraph_, tol) &&
          discreteGraph_.equals(other.discreteGraph_, tol) &&
-         dcGraph_.equals(other.dcGraph_, tol);
+         dcGraph_.equals(other.dcGraph_, tol) &&
+         gaussianGraph_.equals(other.gaussianGraph_, tol);
 }
 
-size_t HybridFactorGraph::size() const {
-  return nonlinearGraph_.size() + discreteGraph_.size() + dcGraph_.size();
+void HybridFactorGraph::clear() {
+  nonlinearGraph_.resize(0);
+  discreteGraph_.resize(0);
+  dcGraph_.resize(0);
+  gaussianGraph_.resize(0);
 }
 
-size_t HybridFactorGraph::size_nonlinear() const {
-  return nonlinearGraph_.size();
+/// Define adding a GaussianFactor to a sum.
+using Sum = DCGaussianMixtureFactor::Sum;
+static Sum& operator+=(Sum& sum, const GaussianFactor::shared_ptr& factor) {
+  using Y = GaussianFactorGraph;
+  auto add = [&factor](const Y& graph) {
+    auto result = graph;
+    result.push_back(factor);
+    return result;
+  };
+  sum = sum.apply(add);
+  return sum;
 }
 
-size_t HybridFactorGraph::size_discrete() const {
-  return discreteGraph_.size();
+Sum HybridFactorGraph::sum() const {
+  if (nrNonlinearFactors()) {
+    throw std::runtime_error(
+        "HybridFactorGraph::sum cannot handle NonlinearFactors.");
+  }
+
+  if (nrDiscreteFactors()) {
+    throw std::runtime_error(
+        "HybridFactorGraph::sum cannot handle DiscreteFactors.");
+  }
+
+  // "sum" all factors, gathering into GaussianFactorGraph
+  DCGaussianMixtureFactor::Sum sum;
+  for (auto&& dcFactor : dcGraph()) {
+    if (auto mixtureFactor =
+            boost::dynamic_pointer_cast<DCGaussianMixtureFactor>(dcFactor)) {
+      sum += *mixtureFactor;
+    } else {
+      throw std::runtime_error(
+          "HybridFactorGraph::sum can only handleDCGaussianMixtureFactors.");
+    }
+  }
+
+  // Add the original factors from the gaussian factor graph
+  for (auto&& gaussianFactor : gaussianGraph()) {
+    sum += gaussianFactor;
+  }
+  return sum;
 }
 
-size_t HybridFactorGraph::size_dc() const { return dcGraph_.size(); }
+std::ostream& operator<<(std::ostream& os,
+                         const GaussianFactorGraph::EliminationResult& er) {
+  os << "ER" << std::endl;
+  return os;
+}
 
-void HybridFactorGraph::clear() {
-  nonlinearGraph_.resize(0);
-  discreteGraph_.resize(0);
-  dcGraph_.resize(0);
+// The function type that does a single elimination step on a variable.
+std::pair<GaussianMixture::shared_ptr, boost::shared_ptr<Factor>>
+EliminateHybrid(const HybridFactorGraph& factors, const Ordering& ordering) {
+  // Create a new decision tree with all factors gathered at leaves.
+  auto sum = factors.sum();
+
+  // Now we need to eliminate each one using conventional Cholesky:
+  // We can use this by creating a *new* decision tree:
+  using GFG = GaussianFactorGraph;
+  using Pair = GaussianFactorGraph::EliminationResult;
```

Comment:
A bit of an ambiguous name but it should be fine since it's only used in the method.